### PR TITLE
[hail] make coalesce and or_else evaluate each argument once

### PR DESCRIPTION
**Pull request by danking**
cc: @tpoterba 

This seems like the correct semantics for these expressions.

### CODE REVIEW

**Comment by tpoterba** [comment]
wait, why? Is there a motivating example? Coalesce and or_else should just be if/else

**Comment by danking** [comment]
@tpoterba Without this change, or_else evaluates the first argument twice:
```
In [4]: print(hl.or_else(hl.int(3) + 4, 5)._ir)
(If (ApplyUnaryPrimOp `!` (IsNA (ApplyBinaryPrimOp `+` (I32 3) (I32 4))))
    (ApplyBinaryPrimOp `+` (I32 3) (I32 4))
    (I32 5))
```

This change makes it:
```
(Let __uid_1 (ApplyBinaryPrimOp `+` (I32 3) (I32 4))
    (If (ApplyUnaryPrimOp `!` (IsNA (Ref __uid_1)))
        (Ref __uid_1)
        (I32 5)))
```

Coalesce is similar:
```
In [6]: print(hl.coalesce(hl.int(1) + 2, hl.int(3) + 4, 5)._ir)
(Let __uid_5 (ApplyBinaryPrimOp `+` (I32 1) (I32 2))
    (If (ApplyUnaryPrimOp `!` (IsNA (Ref __uid_5)))
        (Ref __uid_5)
        (Let __uid_4 (ApplyBinaryPrimOp `+` (I32 3) (I32 4))
            (If (ApplyUnaryPrimOp `!` (IsNA (Ref __uid_4)))
                (Ref __uid_4)
                (I32 5)))))
```

File: hail/python/hail/expr/functions.py [hunk 1]
```diff
@@ -1453,12 +1453,11 @@ def coalesce(*args):
...
-    def make_case(*expr_args):
-        c = case()
-        for e in expr_args:
-            c = c.when(hl.is_defined(e), e)
-        return c.or_missing()
-    return bind(make_case, *exprs)
+    exprs.reverse()
+    coalesced = exprs[0]
+    for expr in exprs[1:]:
+        coalesced = hl.or_else(expr, coalesced)
```
Comment (daniel-goldstein) [inline hail/python/hail/expr/functions.py]: Could this just be done with a fold without having to reverse?

**Comment by tpoterba** [comment]
oh, I see. Good catch!

**Comment by danking** [comment]
@daniel-goldstein sure, functors.reduce with a backwards iterator does the trick. I also added some tests, including one that ensures we do not evaluate arguments after the non-null one.

**Comment by danking** [comment]
@daniel-goldstein buuuuuuump
